### ISSUE TITLE
return type doesn't match compiler inferred type for some functions

### ISSUE THREAD

**Issue by gaika** [JuliaComputing/ArrayFire.jl#92]
```
using Base.Test
f(x) = x .* x
@Test.inferred f(AFArray([3.]))
```
produces

`ERROR: return type ArrayFire.AFArray{Float64,1} does not match inferred return type ArrayFire.AFArray{T,N}`

makes operations with small arrays costlier than they should be as Julia will switch to dynamic dispatch for the return value of this function.

**Comment by ranjanan** [comment]
@gaika Thanks for reporting this. Yes, you're right. This can be improved. This happens because I leave dimension calculation to the runtime. 

One way to fix this is to use a `Base.@pure` function to compute the appropriate dimension of the resultant `AFArray`. I'm playing with that right now. I'll come up with a PR soon.

**Comment by gaika** [comment]
Can this fix  Matrix * Vector (#73) as well?

**Comment by gaika** [comment]
fixed in #121
